src/utilities/include/public.h

```diff
@@ -321,7 +321,7 @@ struct Box {
 /**
  * Boolean operation type: Add (Union), Subtract (Difference), and Intersect.
  */
-enum class OpType { Add, Subtract, Intersect };
+enum class OpType { Add, Subtract, Intersect, KeepP, KeepQ };
```

Comment:
I'd guess we only need one new Op - this API is private anyway. For the public Manifold API, I'm thinking something like `Manifold Manifold::Stamp(const Manifold& stamp, uint32_t originalID) const`, which would apply the given originalID to the portion of the manifold that's inside the stamp manifold.

Reply:
My use case was that if I later came up with a non-manifold operation I wanted to perform, this would give me all of the raw materials I could need in the most computationally efficient way.

The intersection line and re-triangulation would be done on both A and B.  And if appropriate tagging is performed, we would have clear marking of A out of B, A in B, B out of A, and B in A.

Reply:
To be clear, my request here is not asking for non-manifold output from the library.  I am suggesting an API like [`Split()`](https://github.com/elalish/manifold/blob/master/src/manifold/include/manifold.h#L231).  That returns _two_ manifold outputs.

I haven't looked closely at the benchmarking to see if you separate the times required to do the intersection and triangulation vs. the results preparation.  However I really like the idea that these steps are separated and that one could generate multiple results from a single intersection and triangulation (like in `Split()`).  I also like the ability to create consistent triangulations between 'matching' operations.

I don't have a use case for a master BooleanAll function that returns the Union, Intersection, A-B and B-A all in one operation -- but deep down, I really want it and I feel like I'll have a use for it someday....

Reply:
That's fine, but one of my design principals for this library is to not introduce code complexity (especially without full code coverage in our tests) for potential features we haven't determined are useful yet. It's easy enough to come back and add it in later - until then, think about finding a use case.

Reply:
Unlike the BooleanAll, I do have a use case for this one.  It is to allow non-manifold operations by users (outside of the library) without breaking Manifold's design assumption / assurance of manifold only.

This particular example use case also depends on a working thin-manifold solution.  If I was working with a wing with ribs and spars initially defined by planes that extend beyond the surface of the wing.  First, I need to trim the ribs/spars to the wing's surface and keep the part of the ribs/spars inside the wing.  I also need the wing to be 'impressed' wherever the ribs/spars intersect the wing.

Next, I need to intersect all the ribs/spars with each other -- to create connectivity.  But I don't want to discard any of the tris.  I would then iterate over all the 'parts' to mutually intersect them, without throwing away any tris.

Every part would be manifold before, during, and after the calls to Manifold.  They would all end up appropriately trimmed and would have appropriate intersection lines impressed on them.

Finally, I could write them out to a FEM file for application of materials, boundary conditions, and loads, etc.  The FEM model is non-manifold, but all the parts are manifold and Manifold never has to see the non-manifold combination.

Reply:
It also sounds a lot like returning A-B and B-A. As for creating your own connectivity, I have the feeling that won't be as easy as it sounds, particularly because of removal of short edges, which won't happen the same way on the separate outputs. 

For the wing-rib example, I don't believe this would be any different than subtracting the ribs from the wing and then reconnecting the separated parts (same problem as above). Anyway, let's start small and work our way up. You're welcome to create a Sample that demonstrates this kind of use and then we can talk concretely about how much we can improve it with various API changes. For this PR, let's just make a `Stamp` method to keep it easiest to review.